README.md
# MMM-NFL

[![GitHub license](https://img.shields.io/badge/license-MIT-blue.svg?style=flat)](https://raw.githubusercontent.com/fewieden/MMM-NFL/master/LICENSE) ![Build status](https://github.com/fewieden/MMM-NFL/workflows/build/badge.svg) [![Code Climate](https://codeclimate.com/github/fewieden/MMM-NFL/badges/gpa.svg?style=flat)](https://codeclimate.com/github/fewieden/MMM-NFL) [![Known Vulnerabilities](https://snyk.io/test/github/fewieden/mmm-nfl/badge.svg)](https://snyk.io/test/github/fewieden/mmm-nfl)

National Football League Module for MagicMirror<sup>2</sup>

## Example

![](.github/regular-season.png) ![](.github/post-season.png) ![](.github/statistic-modal.png) ![](.github/help-modal.png)

## Dependencies

* An installation of [MagicMirror<sup>2</sup>](https://github.com/MichMich/MagicMirror)
* npm
* [node-fetch](https://www.npmjs.com/package/node-fetch)

## Installation

1. Clone this repo into `~/MagicMirror/modules` directory.
1. Configure your `~/MagicMirror/config/config.js`:

    ```
    {
        module: 'MMM-NFL',
        position: 'top_right',
        config: {
            // all your config options, which are different than their default values
        }
    }
    ```

1. Run command `npm install --production` in `~/MagicMirror/modules/MMM-NFL` directory.

## Global config

| **Option** | **Default** | **Description** |
| --- | --- | --- |
| `locale` | `undefined` | By default it is using your system settings. You can specify the locale in the global MagicMirror config. Possible values are for e.g.: `'en-US'` or `'de-DE'`. |

To set a global config you have to set the value in your config.js file inside the MagicMirror project.

![](.github/global.png)

## Config Options

| **Option** | **Default** | **Description** |
| --- | --- | --- |
| `colored` | `false` | Remove black/white filter of logos. |
| `focus_on` | `false` | Display only matches with teams of this array e.g. `['NYG', 'DAL', 'NE']`. |
| `format` | `'ddd h:mm'` | In which format the date should be displayed. [All Options](http://momentjs.com/docs/#/displaying/format/) |
| `reloadInterval` | `1800000` (30 mins) | How often should the data be fetched. |
| `reverseTeams` | `false` | Flag to switch order of home and away team. |
| `tableSize` | `'small'` | Font size of table. Possible values: `'xsmall'`, `'small'`, `'medium'`, `'large'` and `'xlarge'` |

## OPTIONAL: Voice Control and Modal

This module supports voice control by [MMM-voice](https://github.com/fewieden/MMM-voice) and [MMM-Modal](https://github.com/fewieden/MMM-Modal).
In order to use this feature, it's required to install the voice and modal modules. There are no extra config options for voice control and modals needed.

### Mode

The voice control mode for this module is `FOOTBALL`

### List of all Voice Commands

* OPEN HELP -> Shows the information from the readme here with mode and all commands.
* CLOSE HELP -> Hides the help information.
* SHOW PASSING YARDS STATISTIC -> Shows statistic of players with most passing yards.
* SHOW RUSHING YARDS STATISTIC -> Shows statistic of players with most rushing yards.
* SHOW RECEIVING YARDS STATISTIC -> Shows statistic of players with most receiving yards.
* SHOW TACKLES STATISTIC -> Shows statistic of players with most tackles.
* SHOW SACKS STATISTIC -> Shows statistic of players with most sacks.
* SHOW KICKOFF YARDS STATISTIC -> Shows statistic of players with most kickoff yards.
* SHOW INTERCEPTIONS STATISTIC -> Shows statistic of players with most interceptions.
* SHOW PASSING TOUCHDOWNS STATISTIC -> Shows statistic of players with most passing touchdowns.
* SHOW QUARTERBACK RATING STATISTIC -> Shows statistic of quarterbacks with best rating.
* SHOW RUSHING TOUCHDOWNS STATISTIC -> Shows statistic of players with most rushing touchdowns.
* SHOW RECEPTIONS STATISTIC -> Shows statistic of players with most receptions.
* SHOW RECEIVING TOUCHDOWNS STATISTIC -> Shows statistic of players with most receiving touchdowns.
* SHOW TOTAL POINTS STATISTIC -> Shows statistic of players with most points.
* SHOW TOTAL TOUCHDOWNS STATISTIC -> Shows statistic of players with most touchdowns.
* SHOW PUNT YARDS STATISTIC -> Shows statistic of players with most punt yards.
* SHOW PASSES DEFENDED STATISTIC -> Shows statistic of players with most passes defended.
* HIDE STATISTIC -> Hide statistic informations

## Developer

* `npm run lint` - Lints JS and CSS files.


CHANGELOG.md
# MMM-NFL Changelog

## [Unreleased]

### Fixed

### Added

### Changed

### Removed

## 1.3.1

### Fixed

- [Config option `focus_on` didn't filter match list](https://github.com/fewieden/MMM-NFL/issues/55)

### Removed

- TravisCI integration

## 1.3.0

MagicMirror² version >= 2.15.0 required.

### Added

* Nunjuck templates
* Remote team logos
* Readded live game fetching
* More statistic types
* Integrated MagicMirror logger on server side
* Integrated [MMM-Modal](https://github.com/fewieden/MMM-Modal)
* Implemented suspend, resume and stop behavior
* Added instructions for global config options
* Github actions: `changelog` and `build`
* Github config files

### Changed

* Data structure
* Dimmed bye week
* Project config files
* Dependency updates
* Preview pictures

### Removed

* API provider: NFL XML API
* Local team logos
* Config option: `helmets`

## [1.2.1]

### Added

* Ball possession support for API provider ESPN
* Redzone support for API provider ESPN

### Fixed

* Ball possession team if config option `reverseTeams` is set to `false`

## [1.2.0]

### Added

* New API provider: ESPN

## [1.1.1]

### Fixed

* Bye week translation https://github.com/fewieden/MMM-NFL/issues/37

## [1.1.0]

### Fixed

* Data source
* Installation instructions

### Added

* Config option `tableSize` to specify font size of table.
* Config option `reverseTeams` to invert the display order of home and away team.
* Display of week in header.
* package-lock.json
* Pro bowl mode

### Changed

* Updated icons
* Start in regular mode

### Removed

* team renaming

## [1.0.0]

Initial version


espn.js
const fetch = require('node-fetch');

const seasonStageMapping = {
    1: 'PRE',
    2: 'REG',
    3: 'POST',
    4: 'OFF',
};

const teamNameMapping = {
    LAR: 'LA',
    WSH: 'WAS',
};

const statisticTypes = [
    'passingYards',
    'rushingYards',
    'receivingYards',
    'totalTackles',
    'sacks',
    'kickoffYards',
    'interceptions',
    'passingTouchdowns',
    'quarterbackRating',
    'rushingTouchdowns',
    'receptions',
    'receivingTouchdowns',
    'totalPoints',
    'totalTouchdowns',
    'puntYards',
    'passesDefended'
];

function getFullUrl(apiVersion, path) {
    return `https://site.api.espn.com/apis/site/${apiVersion}/sports/football/nfl${path}`;
}

function getPostGameStatus(period) {
    return period > 4 ? 'final-overtime' : 'final';
}

function getInGameStatus(period) {
    return period > 4 ? 'overtime' : period;
}

function getGameStatus(status = {}) {
    if (status.type?.state === 'pre') {
        return 'pregame';
    } else if (status.type?.name === 'STATUS_HALFTIME') {
        return 'halftime';
    } else if (status.type?.state === 'post') {
        return getPostGameStatus(status.period);
    }

    return getInGameStatus(status.period);
}

function getTeamName(competitor = {}) {
    const team = competitor.team?.abbreviation;

    return teamNameMapping[team] || team;
}

function mapEventEntry(event = {}) {
    const ongoing = !['pre', 'post'].includes(event.status?.type?.state);

    const possessionTeamId = event.competitions?.[0]?.situation?.possession;
    const possessionTeam = event.competitions?.[0]?.competitors?.find(c => c.id === possessionTeamId);

    return {
        timestamp: event.date,
        status: getGameStatus(event.status),
        remainingTime: ongoing && event.status?.displayClock,
        ballPossession: getTeamName(possessionTeam),
        inRedZone: event.competitions?.[0]?.situation?.isRedZone,
        homeTeam: getTeamName(event.competitions?.[0]?.competitors?.[0]),
        homeScore: event.competitions?.[0]?.competitors?.[0]?.score,
        homeLogo: event.competitions?.[0]?.competitors?.[0]?.team?.logo,
        awayTeam: getTeamName(event.competitions?.[0]?.competitors?.[1]),
        awayScore: event.competitions?.[0]?.competitors?.[1]?.score,
        awayLogo: event.competitions?.[0]?.competitors?.[1]?.team?.logo
    };
}

async function getData() {
    const response = await fetch(getFullUrl('v2', '/scoreboard'));

    if (!response.ok) {
        throw new Error('failed to fetch scoreboard');
    }

    const parsedResponse = await response.json();

    const details = {
        week: parsedResponse?.week?.number,
        season: parsedResponse?.season?.year,
        stage: seasonStageMapping[parsedResponse?.season?.type]
    };

    const events = parsedResponse?.events || [];

    const scores = events.map(mapEventEntry).sort((a, b) => {
        if (a.timestamp === b.timestamp) {
            return 0;
        }

        return a.timestamp > b.timestamp ? 1 : -1
    });

    return { details, scores };
}

function mapPlayerEntry(player = {}) {
    return {
        value: player.displayValue,
        name: player.athlete.fullName,
        avatar: player.athlete.headshot.href,
        team: getTeamName(player),
        logo: player?.team?.logos?.[0]?.href,
    };
}

async function getStatistics(type) {
    if (!statisticTypes.includes(type)) {
        throw new Error(`Unsupported statistic type: ${type}`);
    }

    const response = await fetch(getFullUrl('v3', '/leaders'));

    if (!response.ok) {
        throw new Error('failed to fetch scoreboard');
    }

    const parsedResponse = await response.json();

    const category = parsedResponse?.leaders?.categories?.find(c => c.name === type);

    const players = category?.leaders || [];

    const leaders = players.map(mapPlayerEntry);

    return leaders;
}

module.exports = { getData, getStatistics };


node_helper.js
/* Magic Mirror
 * Module: MMM-NFL
 *
 * By fewieden https://github.com/fewieden/MMM-NFL
 * MIT Licensed.
 */

/* eslint-env node */

const NodeHelper = require('node_helper');
const Log = require('logger');

const ESPN = require('./espn');

const ONE_MINUTE = 60 * 1000;

const commandStatisticMapping = {
    passingTouchdowns: ['PASSING', 'TOUCHDOWNS'],
    rushingTouchdowns: ['RUSHING', 'TOUCHDOWNS'],
    receivingTouchdowns: ['RECEIVING', 'TOUCHDOWNS'],
    totalTouchdowns: ['TOTAL', 'TOUCHDOWNS'],
    passingYards: ['PASSING', 'YARDS'],
    rushingYards: ['RUSHING', 'YARDS'],
    receivingYards: ['RECEIVING', 'YARDS'],
    sacks: ['SACKS'],
    interceptions: ['INTERCEPTIONS'],
    totalTackles: ['TACKLES'],
    quarterbackRating: ['QUARTERBACK', 'RATING'],
    receptions: ['RECEPTIONS'],
    passesDefended: ['PASSES', 'DEFENDED'],
    totalPoints: ['TOTAL', 'POINTS'],
    puntYards: ['PUNT', 'YARDS'],
    kickoffYards: ['KICKOFF', 'YARDS']
};

module.exports = NodeHelper.create({
    requiresVersion: '2.15.0',

    scores: [],
    reloadInterval: null,
    liveInterval: null,

    async socketNotificationReceived(notification, payload) {
        if (notification === 'CONFIG') {
            this.config = payload;
            this.reloadInterval = setInterval(() => {
                this.getData();
            }, this.config.reloadInterval);
            this.liveInterval = setInterval(() => {
                this.fetchOnLiveState();
            }, ONE_MINUTE);
            await this.getData();
        } else if (notification === 'VOICE_COMMAND') {
            await this.handleVoiceCommand(payload);
        } else if (notification === 'SUSPEND') {
            this.stop();
        }
    },

    async getData() {
        try {
            const data = await ESPN.getData();
            this.scores = data.scores;
            this.sendSocketNotification('SCORES', data);
        } catch (error) {
            Log.error(`Error getting NFL scores ${error}`);
        }
    },

    async getStatisticsFromVoiceCommand(command) {
        try {
            let type = Object.keys(commandStatisticMapping)[0];
            for (const statisticType in commandStatisticMapping) {
                const matching = commandStatisticMapping[statisticType].every(word => command.includes(word));
                if (matching) {
                    type = statisticType;
                    break;
                }
            }
            const statistics = await ESPN.getStatistics(type);
            this.sendSocketNotification('STATISTICS', { type, statistics });
        } catch (error) {
            Log.error(`Error getting NFL statistics ${error}`);
        }
    },

    fetchOnLiveState() {
        const currentTime = new Date().toISOString();

        const endStates = ['final', 'final-overtime'];
        const liveMatch = this.scores.find(match => currentTime > match.timestamp && !endStates.includes(match.status));

        if (liveMatch) {
            this.getData();
        }
    },

    shouldCloseOpenModal(command) {
        return command.includes('HELP') && command.includes('CLOSE') || command.includes('STATISTIC') && command.includes('HIDE');
    },

    shouldShowHelpModal(command) {
        return command.includes('HELP') && command.includes('OPEN');
    },

    shouldShowStatisticModal(command) {
        return command.includes('SHOW') && command.includes('STATISTIC');
    },

    async handleVoiceCommand(command = '') {
        if (this.shouldCloseOpenModal(command)) {
            this.sendSocketNotification('CLOSE_MODAL');
        } else if (this.shouldShowHelpModal(command)) {
            this.sendSocketNotification('OPEN_HELP_MODAL');
        } else if (this.shouldShowStatisticModal(command)) {
            await this.getStatisticsFromVoiceCommand(command);
        }
    },

    stop() {
        clearInterval(this.liveInterval);
        clearInterval(this.reloadInterval);
    }
});


StatisticsAPI.js
/* Magic Mirror
 * Module: MMM-NFL
 *
 * By fewieden https://github.com/fewieden/MMM-NFL
 * MIT Licensed.
 */

/* eslint-env node */

const jsdom = require('jsdom');

module.exports = {

    URL: 'http://www.espn.com/nfl/statistics',
    CACHE_TIME: 2 * 60 * 1000, // 2 minutes

    statistics: {
        timestamp: null,
        data: {}
    },

    getStats(type, callback) {
        const time = Date.now();
        if (!this.statistics.timestamp || this.statistics.timestamp < time - this.CACHE_TIME ||
            !Object.prototype.hasOwnProperty.call(this.statistics.data, type.toUpperCase())) {
            jsdom.env({
                url: this.URL,
                done: (err, window) => {
                    if (err) {
                        callback(err, null);
                    } else {
                        const statistics = {};
                        const statisticsWrapper = window.document.querySelector('#my-players-table');
                        if (statisticsWrapper) {
                            const teams = [statisticsWrapper.children[0], statisticsWrapper.children[1]];
                            for (let i = 0; i < teams.length; i += 1) {
                                const types = teams[i].children;
                                for (let n = 0; n < types.length; n += 1) {
                                    const rows = types[n].querySelector('tbody').children;
                                    if (rows) {
                                        statistics[rows[0].children[0].textContent] = {
                                            unit: rows[0].children[1].textContent,
                                            players: []
                                        };
                                        for (let x = 1; x < rows.length - 1; x += 1) {
                                            const text = rows[x].children[rows[x].children.length - 2].textContent;
                                            statistics[rows[0].children[0].textContent].players.push({
                                                position: text[0],
                                                player: text.slice(3, text.lastIndexOf(',')),
                                                team: text.slice(text.lastIndexOf(',') + 2) === 'WSH' ? 'WAS' : text.slice(text.lastIndexOf(',') + 2),
                                                value: rows[x].children[rows[x].children.length - 1].textContent
                                            });
                                        }
                                    }
                                }
                            }

                            window.close();

                            this.statistics = {
                                timestamp: time,
                                data: statistics
                            };

                            if (Object.prototype.hasOwnProperty.call(this.statistics.data, type.toUpperCase())) {
                                callback(null, { type, data: this.statistics.data[type.toUpperCase()] });
                            } else {
                                callback(`Statistics for ${type} not found!`, null);
                            }
                        } else {
                            callback('Source changed site layout!', null);
                        }
                    }
                }
            });
        } else {
            callback(null, { type, data: this.statistics.data[type.toUpperCase()] });
        }
    }
};


MMM-NFL.js
/* global Module Log moment config */

/* Magic Mirror
 * Module: MMM-NFL
 *
 * By fewieden https://github.com/fewieden/MMM-NFL
 * MIT Licensed.
 */

Module.register('MMM-NFL', {
    modes: {
        PRE: 'Preseason',
        REG: 'Regular-Season',
        POST: 'Post-Season',
        OFF: 'Offseason',
    },

    details: {
        season: new Date().getFullYear(),
        stage: 'REG'
    },

    states: {
        1: '1ST_QUARTER',
        2: '2ND_QUARTER',
        3: '3RD_QUARTER',
        4: '4TH_QUARTER',
        halftime: 'HALF_TIME',
        overtime: 'OVER_TIME',
        final: 'FINAL',
        'final-overtime': 'FINAL_OVERTIME',
        pregame: 'UPCOMING'
    },

    defaults: {
        colored: false,
        focus_on: false,
        format: 'ddd h:mm',
        reloadInterval: 30 * 60 * 1000, // every 30 minutes
        reverseTeams: false,
        tableSize: 'small'
    },

    getTranslations() {
        return {
            en: 'translations/en.json',
            de: 'translations/de.json'
        };
    },

    getScripts() {
        return ['moment.js'];
    },

    getStyles() {
        return ['font-awesome.css', 'MMM-NFL.css'];
    },

    getTemplate() {
        return `templates/${this.name}.njk`;
    },

    findTeamInScores(team) {
        return this.scores.find(m => team === m.homeTeam || team === m.awayTeam);
    },

    getFocusedTeamsWithByeWeeks() {
        if (!Array.isArray(this.config.focus_on) || !Array.isArray(this.scores)) {
            return [];
        }

        return this.config.focus_on.filter(team => !this.findTeamInScores(team));
    },

    getFilteredScores() {
        if (!Array.isArray(this.config.focus_on) || !Array.isArray(this.scores)) {
            return this.scores;
        }

        return this.scores.filter(m => this.config.focus_on.includes(m.homeTeam) || this.config.focus_on.includes(m.awayTeam));
    },

    getTemplateData() {
        return {
            states: this.states,
            modes: this.mode,
            details: this.details,
            config: this.config,
            scores: this.getFilteredScores(),
            focusedTeamsWithByeWeeks: this.getFocusedTeamsWithByeWeeks()
        };
    },

    getVoiceData() {
        return {
            mode: 'FOOTBALL',
            sentences: [
                'OPEN HELP',
                'CLOSE HELP',
                'SHOW PASSING YARDS STATISTIC',
                'SHOW RUSHING YARDS STATISTIC',
                'SHOW RECEIVING YARDS STATISTIC',
                'SHOW TACKLES STATISTIC',
                'SHOW SACKS STATISTIC',
                'SHOW KICKOFF YARDS STATISTIC',
                'SHOW INTERCEPTIONS STATISTIC',
                'SHOW PASSING TOUCHDOWNS STATISTIC',
                'SHOW QUARTERBACK RATING STATISTIC',
                'SHOW RUSHING TOUCHDOWNS STATISTIC',
                'SHOW RECEPTIONS STATISTIC',
                'SHOW RECEIVING TOUCHDOWNS STATISTIC',
                'SHOW TOTAL POINTS STATISTIC',
                'SHOW TOTAL TOUCHDOWNS STATISTIC',
                'SHOW PUNT YARDS STATISTIC',
                'SHOW PASSES DEFENDED STATISTIC',
                'HIDE STATISTIC'
            ]
        };
    },

    start() {
        Log.info(`Starting module: ${this.name}`);
        this.addGlobals();
        this.addFilters();
        this.sendSocketNotification('CONFIG', this.config);
        moment.locale(config.locale);
    },

    suspend() {
        this.sendSocketNotification('SUSPEND', this.config);
    },

    resume() {
        this.sendSocketNotification('CONFIG', this.config);
    },

    notificationReceived(notification, payload, sender) {
        if (notification === 'ALL_MODULES_STARTED') {
            this.sendNotification('REGISTER_VOICE_MODULE', this.getVoiceData());
        } else if (notification === 'VOICE_FOOTBALL' && sender.name === 'MMM-voice') {
            this.sendSocketNotification('VOICE_COMMAND', payload);
        } else if (notification === 'VOICE_MODE_CHANGED' && sender.name === 'MMM-voice' && payload.old === this.getVoiceData().mode) {
            this.sendNotification('CLOSE_MODAL');
        }
    },

    openStatisticsModal({ type, statistics }) {
        this.sendNotification('OPEN_MODAL', {
            template: 'templates/StatisticsModal.njk',
            data: {
                type,
                statistics,
                config: this.config,
                fns: { translate: this.translate.bind(this) }
            }
        });
    },

    openHelpModal() {
        this.sendNotification('OPEN_MODAL', {
            template: 'templates/HelpModal.njk',
            data: {
                ...this.getVoiceData(),
                fns: { translate: this.translate.bind(this) }
            }
        });
    },

    socketNotificationReceived(notification, payload) {
        if (notification === 'SCORES') {
            this.scores = payload.scores;
            this.details = payload.details;
            this.updateDom(300);
        } else if (notification === 'STATISTICS') {
            this.openStatisticsModal(payload);
        } else if (notification === 'OPEN_HELP_MODAL') {
            this.openHelpModal();
        } else if (notification === 'CLOSE_MODAL') {
            this.sendNotification('CLOSE_MODAL');
        }
    },

    addGlobals() {
        this.nunjucksEnvironment().addGlobal('includes', (array, item) => array.includes(item));
    },

    addFilters() {
        this.nunjucksEnvironment().addFilter('formatDate', timestamp => moment(timestamp).format(this.config.format));

        this.nunjucksEnvironment().addFilter('iconUrl', teamName => this.file(`icons/${teamName}${this.config.helmets ? '_helmet' : ''}.png`));
    }
});
